File: TheepotMobile/TheepotMobile/Models/Vakantieplanningen.swift
import Foundation

/// Tabel `vakantie_planningen`.
struct VakantiePlanning: Codable, Identifiable, Hashable {
    let id: String
    let naam: String
    var vakantie: String
    var thema: String?
    let startDatum: String
    let eindDatum: String
    var gepubliceerd: Bool
    var startDatumNoord: String?
    var eindDatumNoord: String?

    enum CodingKeys: String, CodingKey {
        case id, naam, vakantie, thema
        case startDatum = "start_datum"
        case eindDatum = "eind_datum"
        case gepubliceerd
        case startDatumNoord = "start_datum_noord"
        case eindDatumNoord = "eind_datum_noord"
    }
}

/// Tabel `vakantie_weken`.
struct VakantieWeek: Codable, Identifiable, Hashable {
    let id: String
    let planningId: String
    let weekNummer: Int
    let naam: String

    enum CodingKeys: String, CodingKey {
        case id
        case planningId = "planning_id"
        case weekNummer = "week_nummer"
        case naam
    }
}

/// Tabel `vakantie_activiteiten`.
struct VakantieActiviteit: Codable, Identifiable, Hashable {
    let id: String
    let weekId: String
    let dag: Dag
    let volgorde: Int
    let categorie: String
    let naam: String
    let beschrijving: String?
    let benodigdheden: [String]?
    let activiteitId: String?
    var afbeeldingPad: String?

    enum CodingKeys: String, CodingKey {
        case id
        case weekId = "week_id"
        case dag, volgorde, categorie, naam, beschrijving, benodigdheden
        case activiteitId = "activiteit_id"
        case afbeeldingPad = "afbeelding_pad"
    }
}

/// Tabel `activiteiten` — de activiteitenbibliotheek. Alleen gebruikt om een
/// bestaande activiteit in een vakantieplanning te kunnen zetten; geen eigen
/// scherm in de iOS-app.
struct BibliotheekActiviteit: Decodable, Identifiable, Hashable {
    let id: String
    let naam: String
    let categorie: String
    let thema: [String]
    let tijdsduur: Int?
    let materialen: [String]?
    let beschrijving: String?

    /// `thema` staat in de database soms als losse tekst, soms als array
    /// (zie app/activiteiten/page.tsx) — hier altijd genormaliseerd tot een lijst.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        naam = try c.decode(String.self, forKey: .naam)
        categorie = try c.decode(String.self, forKey: .categorie)
        tijdsduur = try c.decodeIfPresent(Int.self, forKey: .tijdsduur)
        materialen = try c.decodeIfPresent([String].self, forKey: .materialen)
        beschrijving = try c.decodeIfPresent(String.self, forKey: .beschrijving)

        if let lijst = try? c.decodeIfPresent([String].self, forKey: .thema) {
            thema = lijst
        } else if let tekst = try? c.decodeIfPresent(String.self, forKey: .thema) {
            thema = [tekst]
        } else {
            thema = []
        }
    }

    enum CodingKeys: String, CodingKey {
        case id, naam, categorie, thema, tijdsduur, materialen, beschrijving
    }
}

File: TheepotAndroid/app/src/main/kotlin/nl/bsodetheepot/mobile/data/models/Vakantieplanningen.kt
package nl.bsodetheepot.mobile.data.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Tabel `vakantie_planningen`. */
@Serializable
data class VakantiePlanning(
    val id: String,
    val naam: String,
    val vakantie: String,
    val thema: String? = null,
    @SerialName("start_datum") val startDatum: String,
    @SerialName("eind_datum") val eindDatum: String,
    val gepubliceerd: Boolean = false,
    @SerialName("start_datum_noord") val startDatumNoord: String? = null,
    @SerialName("eind_datum_noord") val eindDatumNoord: String? = null,
)

/** Tabel `vakantie_weken`. */
@Serializable
data class VakantieWeek(
    val id: String,
    @SerialName("planning_id") val planningId: String,
    @SerialName("week_nummer") val weekNummer: Int,
    val naam: String,
)

/** Tabel `vakantie_activiteiten`. */
@Serializable
data class VakantieActiviteit(
    val id: String,
    @SerialName("week_id") val weekId: String,
    val dag: Dag,
    val volgorde: Int = 0,
    val categorie: String? = null,
    val naam: String,
    val beschrijving: String? = null,
    val benodigdheden: List<String>? = null,
    @SerialName("activiteit_id") val activiteitId: String? = null,
    @SerialName("afbeelding_pad") val afbeeldingPad: String? = null,
)
